Build system: use ocaml{c,opt}.opt instead of ocaml{c,opt} when available

gasche opened the pull request:
This PR supersedes #8839, it tweaks the build system to use native-compiled compilers (and lexer) in most of the tools that can be compiled after the .opt compilers are built in the opt.opt (world.opt) build scenarios: otherlibs, ocamltest, ocamldoc and the debugger.

On my machine the total `world.opt` build time with `-j5` decreases from 3m30s to 2m50s, and it seems to save around a minute of CPU time.

gasche committed:
ocamldoc Makefile: use .opt versions of the build tools when available

On my machine, a sequential build of (make all allopt) goes from 31s
to 11s, and a parallel build (-j5) goes from 16s to 6s.

gasche committed:
Changes

gasche committed:
ocamltest Makefile: use ocaml{c,opt}.opt when available

On my machine the sequential build time goes from 7.5s to 3.5s.

gasche committed:
debugger Makefile: use ocamlc.opt when available

On my machine the sequential build time goes from 3.7s to 1.3s

gasche committed:
otherlibs Makefiles: use ocaml{c,opt}.opt when available

On my machine, the sequential build time goes from 30s to 10s.

gasche committed:
Makefile.best_binaries: best-effort commands for OCAML{C,OPT,DEP,LEX}

pmetzger commented on Makefile:
```diff
@@ -422,8 +422,8 @@ opt.opt: checknative
 	$(MAKE) ocamlc.opt
-	$(MAKE) otherlibraries $(WITH_DEBUGGER) $(WITH_OCAMLDOC) ocamltest
 	$(MAKE) ocamlopt.opt
+	$(MAKE) otherlibraries $(WITH_DEBUGGER) $(WITH_OCAMLDOC) ocamltest
```
This is nearly the simplest fix possible to produce such a useful result.

pmetzger commented:
Very nice fix!

gasche commented:
If `ocamlc.opt` is older than `ocamlc` then it should not be used silently, because it may be stale (`make world.opt`, then hack on the compiler, then just `make core`, then `make ocamldoc`, the changes to the compiler should be taken into account. In that case I plan to emit a warning and revert to the most-recently-built tool (the bytecode one).

The warning cannot be observed in a cold-build scenario, and it should not be observable when running `make world.opt` several times, but it could be observed if rebuilding from an existing build (so mostly during development). It will mention the two ways to silence the warning, either to remove the stale native tools, or to `touch` them to use them.

gasche committed:
Makefile.best_binaries: only use the native tool when it is newer (not stale)

gasche commented:
I have implemented the newer-than check (it's a Make function rather than a shell script to easily make use of Make's `$(info ...)` printing function, but Make functions are not very pleasant to write or read). A warning such as the following is displayed if, for example, ocamlc is newer than ocamlc.opt:

> Warning: we are not using the native binary ocamlc.opt because it is older than the bytecode binary ocamlc; you should silence this warning by either removing ocamlc.opt or rebuilding it (or `touch`-ing it) if you want it used.

gasche committed:
Makefile: minor fix to the ocamltest rules

(make -C ocamltest) will run the first rule of the ocamltest Makefile,
which happens to be the 'all' rule. This is not robust to the addition
of new rules in some of the included Makefile, and indeed a previous
version of the previous commit broke it inadvertently.

xavierleroy commented:
> I have implemented the newer-than check (it's a Make function rather than a shell script to easily make use of Make's $(info ...) printing function, but Make functions are not very pleasant to write or read). 

I applaud the effort (of using make functions) but dislike the smells of scripting (using `eval`, passing variables by reference, escaping `$`).  Here is a more functional version that almost looks like Lisp code:

```
check_not_stale = \
  $(if $(shell test $(ROOTDIR)/$1 -nt $(ROOTDIR)/$2 && echo stale), \
    $(info Warning: we are not using the native binary $2 \
because it is older than the bytecode binary $1; \
you should silence this warning by either removing $2 \
or rebuilding it (or `touch`-ing it) if you want it used.), \
    ok)

choose_best = \
  $(if $(and $(wildcard $(ROOTDIR)/$2), $(strip $(call check_not_stale,$1,$2))), \
    $(ROOTDIR)/$2, \
    $(ROOTDIR)/$1)

BEST_OCAMLC := $(call choose_best,ocamlc,ocamlc.opt)
BEST_OCAMLOPT := $(call choose_best,ocamlopt,ocamlopt.opt)
BEST_OCAMLLEX := $(call choose_best,lex/ocamllex,lex/ocamllex.opt)

```

xavierleroy committed:
nicer style in Makefile.best_binaries

gasche commented:
Very nice! I pushed a patch with your version. (In the bytecode case we need to prefix with `$(CAMLRUN)`.)

I had thought of doing something like that (without the assignment) but I couldn't find a way that would (1) statically make the choice at Makefile-invocation time (I'm worried of the performance cost of checking at each compiler invocation) and (2) keep using the `=` assignment that allows other Makefiles to redefine `CAMLRUN` and get away with it. This is one reason why I used `eval`, which allows to statically compute the code to run dynamically -- staged programming!

In fact I checked, and CAMLRUN is not redefined anywhere in this style (some other variables use redefinitions and lazy expansions, in particular the CAML{C,OPT} and CAML{C,OPT}_CMD pairing) -- it is sometimes redefined on recursive invocation of Make (to promote, for example) but that should be fine. So we don't need to keep (2) and can use eager assignment with your version.

dra27 commented on Makefile.best_binaries:
```diff
@@ -0,0 +1,45 @@
+    ok)
+
+choose_best = \
+  $(if $(and $(wildcard $(ROOTDIR)/$2), $(strip $(call check_not_stale,$1,$2))), \
```
The formatting is introducing spaces at the start of `BEST_CAMLC` which is what's breaking the Windows build. You either have to put this on one big line, and really annoy @shindere, or do this:

```make
  $(if $(and $(wildcard $(ROOTDIR)/$2), $(strip \
  $(call check_not_stale,$1,$2))),$(ROOTDIR)/$2,$(CAMLRUN) $(ROOTDIR)/$1)
```
which keeps the lines under 80 chars, but breaks in a less readable place, or do this piece of GNU make hackery...

```make
  $(if $(and $(wildcard $(ROOTDIR)/$2), $(strip $(call check_not_stale,$1,$2))),$\
    $(ROOTDIR)/$2,$\
    $(CAMLRUN) $(ROOTDIR)/$1)
```

(or you have to put some Windows special case stuff in to convert the forward slashes to back slashes)

Sorry about that 🙁

gasche replied:
What about just adding a call to `strip`? Would the following work?

```
choose_best = $(strip $(if \
   $(and $(wildcard $(ROOTDIR)/$2),$(strip $(call check_not_stale,$1,$2))), \
    $(ROOTDIR)/$2, \
    $(CAMLRUN) $(ROOTDIR)/$1)))
```

gasche replied:
(I just pushed the `strip` version for CI golf.)

dra27 replied:
Hah! Too long a day, clearly 🙂

gasche commented:
(cc @shindere this is the other Makefile-oriented PR that I mentioned.)

shindere commented:
I am wondering why e.g. `choose_best` needs to take two arguemnts, given
that, as far as I can tell, it will always be called with `foo` and
`foo.opt`. would it be okay to rename it to, say, `find_tool` and to
give it only one argument? I am suspecting that this could be used to
choose between `ocamldoc` and `ocmaldoc.opt` too, right?

gasche commented:
I addressed the review comments of @shindere -- I have a one-argument instead of two-argument function now. `ocamldoc` is not using this mechanism because it is not generic enough -- the invocation command for `ocamldoc` is fairly complex due to dynamic linking.

shindere commented:
I'm not that found of the proliferation of makefiles, even less found of
it when it's at the top-level of the source tree.

Why not adding these definitions to `Makefile.common` directly,
actually?

dra27 commented:
I think it's good to have a mechanism where these definitions must be explicitly opted into. The alternative would be to have them in `Makefile.common` but have a variable which can be defined before `Makefile.common` is `include`d to enable them.

Personally I think having the `Makefile.best_binaries` as it is better

dra27 commented on Makefile.best_binaries:
```diff
@@ -0,0 +1,45 @@
+    ok)
+
+choose_best = $(strip $(if \
+   $(and $(wildcard $(ROOTDIR)/$1.opt),$(strip $(call check_not_stale,$1,$1.opt))), \
```
Alas, this line is failing `check-typo` (I can't remember if you're a fan of the githook or not - worrying that if you are, that it's not working??)

dbuenzli replied:
@dra27 could we have instructions in `HACKING.adoc` on how to enable them (instructions were given in a discussion not long time ago) ? It's quite annoying to make PRs have the slow CI fail only on this.

dra27 replied:
They've been in `CONTRIBUTING.md` since the hook was added (L71-L72)

dbuenzli replied:
Ah thanks ! Maybe there are too much of these documents...

shindere commented:
David Allsopp (2019/09/27 01:45 -0700):
> I think it's good to have a mechanism where these definitions must be
> explicitly opted into.

Why?

> The alternative would be to have them in `Makefile.common` but have a
> variable which can be defined before `Makefile.common` is `include`d
> to enable them.

If anything, I'd prefer that, although I'd not like it very much either.

> Personally I think having the `Makefile.best_binaries` as it is better

It does not really scale...

shindere commented:
Daniel Bünzli (2019/09/27 01:59 -0700):
> dbuenzli commented on this pull request.
>
>
>
> > +
> +# It exports definitions of BEST_OCAML{C,OPT,LEX,DEP} commands that
> +# run to either the bytecode binary built in the repository or the
> +# native binary, if available. Note that they never use the boot/
> +# versions: we assume that ocamlc, ocamlopt, etc. have been run first.
> +
> +check_not_stale = \
> +  $(if $(shell test $(ROOTDIR)/$1 -nt $(ROOTDIR)/$2 && echo stale), \
> +    $(info Warning: we are not using the native binary $2 \
> +because it is older than the bytecode binary $1; \
> +you should silence this warning by either removing $2 \
> +or rebuilding it (or `touch`-ing it) if you want it used.), \
> +    ok)
> +
> +choose_best = $(strip $(if \
> +   $(and $(wildcard $(ROOTDIR)/$1.opt),$(strip $(call check_not_stale,$1,$1.opt))), \
>
> @dra27 could we have instructions in `HACKING.adoc` on how to enable
> them (instructions were given a discussion not long time ago) ? It's
> quite annoying to make PRs have the slow CI fail only on this.

There's even a proposal to include invoking `check-typo` in the
compiler's build process on dev builds. That might even be better and
does not exclude the pre-commit hook anyway.

shindere commented:
Daniel Bünzli (2019/09/27 02:26 -0700):
> Ah thanks ! Maybe there are too much of these documents...

Both files seem already a bit large, but I agree that it would be nice
to have less of them.

gasche commented:
> > Personally I think having the `Makefile.best_binaries` as it is better
>
> It does not really scale...

I'm not sure what sort of scalability you have in mind here, is it the scalability of `ls` at the root? When we write programs, we tend to consider that it is good style to split distinct part of the codebase in different files, instead of having everything in a single source file.

We could think of moving the sub-Makefiles into `tools/` for example, or a new `makefiles/` directory (as we used to have in the testsuite).

shindere commented:
Gabriel Scherer (2019/09/27 07:16 -0700):
> I'm not sure what sort of scalability you have in mind here, is it the
> scalability of `ls` at the root?

Of course not :)

> When we write programs, we tend to consider that it is good style to split distinct part of the codebase in different files, instead of having everything in a single source file.
>
> We could think of moving the sub-Makefiles into `tools/` for example,
> or a new `makefiles/` directory (as we used to have in the testsuite).

Sure, a directory with the sub-makefiles owuld be better, I think. Still
I don't like it too much to have one new Makefile for each thing we may
invent. To me it feels error-prone to multiply the makefiles because
there is always the risk to forget to include one, or to include it in
the wrong place, stuff like that.

gasche commented:
I don't see the problem. If you forget to include a makefile that provides definitions/variables, then your makefile code fails and you quickly notice the problem. On the other hand, having a monster Makefile that defines everything at once makes it harder to reason about, for example, which definitions are or are not in scope at a given point. (For example, everytime you add a definition there, you have to check all Makefiles including it, which is all makefiles, for naming conflicts.)

shindere commented:
Well `Makefile.common` is far from being a monster at the moment, IMO,
but if everybody agrees, fine.

gasche commented:
Could we try to make faster progress on this PR? Saving 40s to a minute of compiler full-build time seems like an important quality-of-life improvement for developers (and could also help CI response times), it's disappointing to wait for two months to make a decision.

dra27 reviewed (approved):
I think from the conversation @shindere was happy for the `Makefile` layout to stay as-is - but that can of course be addressed in a subsequent PR

gasche commented:
Thanks! (I'm sorry for my sprout of grumpiness now.)

I had just rebased the PR to fix the check-typo issue, so we should wait for CI.

dra27 commented:
I took no grumpiness 🙂 - feel free to merge when AppVeyor comes back

dra27 commented:
I presume as you rebased you'd like to keep the commit history?

gasche commented:
I just merged, thanks everyone for the feedback.
